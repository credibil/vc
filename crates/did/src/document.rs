//! # DID Document
//!
//! A DID Document is a JSON-LD document that contains information related to a DID.

use std::collections::HashMap;
use std::fmt::{self, Display, Formatter};

use chrono::{DateTime, Utc};
use core_utils::{Kind, Quota};
use proof::jose::jwk::PublicKeyJwk;
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::did;

/// DID resolution functions required to be implemented by conforming DID resolvers.
pub trait Operator {
    /// Creates a DID document for the given DID method.
    ///
    /// # Errors
    ///
    /// Returns an error if the DID Document cannot be created.
    fn create(&self, did: &str, options: CreateOptions) -> did::Result<Document>;

    /// Reads a DID document for the given DID method.
    #[allow(dead_code)]
    fn read(&self, did: &str, options: CreateOptions) -> did::Result<Document>;

    /// Updates a DID document for the given DID method.
    #[allow(dead_code)]
    fn update(&self, did: &str, options: CreateOptions) -> did::Result<Document>;

    /// Deactivates a DID document for the given DID method.
    #[allow(dead_code)]
    fn deactivate(&self, did: &str, options: CreateOptions) -> did::Result<()>;
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateOptions {
    /// Format to use for the the public key.
    pub public_key_format: PublicKeyFormat,

    /// Default context for the DID document. SHOULD be set to
    /// `"https://www.w3.org/ns/did/v1"`.
    pub default_context: String,

    /// Enable experimental public key types. SHOULD be set to "false".
    pub enable_experimental_public_key_types: bool,

    /// Will add a `keyAgreement` object to the DID document.
    pub enable_encryption_key_derivation: bool,

    // service_endpoints: Vec<Value>,
    // verification_methods: Vec<Value>,
    // authentication: Vec<Value>,
    /// Additional options.
    #[serde(flatten)]
    pub additional: Option<HashMap<String, String>>,
}

impl Default for CreateOptions {
    fn default() -> Self {
        Self {
            public_key_format: PublicKeyFormat::default(),
            enable_experimental_public_key_types: false,
            default_context: "https://www.w3.org/ns/did/v1".to_string(),
            enable_encryption_key_derivation: false,
            additional: None,
        }
    }
}

#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct Document {
    #[serde(rename = "@context")]
    pub context: Vec<Kind<Value>>,

    /// The DID for a particular DID subject.
    ///
    /// The subject is defined as the entity identified by the DID and described by the
    /// DID document. Anything can be a DID subject: person, group, organization,
    /// physical thing, digital thing, logical thing, etc.
    pub id: String,

    /// A set of URIs taht are other identifiers for the subject of the above DID.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub also_known_as: Option<Vec<String>>,

    /// One or more strings that conform to the rules DID Syntax. The corresponding
    /// DID document(s) SHOULD contain verification relationships that explicitly
    /// permit the use of certain verification methods for specific purposes.
    ///
    /// Any verification methods contained in the related DID documents
    /// SHOULD be accepted as authoritative, such that proofs that satisfy those
    /// verification methods are to be considered equivalent to proofs provided by the
    /// DID subject.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub controller: Option<Quota<String>>,

    /// A set of services, that express ways of communicating with the DID subject
    /// or related entities.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub service: Option<Vec<Service>>,

    /// If set, MUST be a set of verification methods for the DID subject.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub verification_method: Option<Vec<VerificationMethod>>,

    /// -----------------------------------------------------------------------
    /// Verification Relationships
    ///
    /// A verification relationship expresses the relationship between the DID subject
    /// and a verification method.
    /// -----------------------------------------------------------------------

    /// The `authentication` verification relationship is used to specify how the DID
    /// subject is expected to be authenticated, for purposes such as logging into
    /// a website or engaging in any sort of challenge-response protocol.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub authentication: Option<Vec<Kind<VerificationMethod>>>,

    /// The `assertion_method` verification relationship is used to specify how the DID
    /// subject is expected to express claims, such as for the purposes of issuing a
    /// Verifiable Credential.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub assertion_method: Option<Vec<Kind<VerificationMethod>>>,

    /// The `capability_invocation` verification relationship is used to specify a
    /// verification method that might be used by the DID subject to invoke a
    /// cryptographic capability, such as the authorization to update the DID Document.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub capability_invocation: Option<Vec<Kind<VerificationMethod>>>,

    /// The `capability_delegation` verification relationship is used to specify a
    /// mechanism that might be used by the DID subject to delegate a cryptographic
    /// capability to another party, such as delegating the authority to access a
    /// specific HTTP API to a subordinate.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub capability_delegation: Option<Vec<Kind<VerificationMethod>>>,

    /// The `key_agreement` verification relationship is used to specify how an entity
    /// can generate encryption material in order to transmit confidential information
    /// intended for the DID subject, such as for the purposes of establishing a secure
    /// communication channel with the recipient.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub key_agreement: Option<Vec<Kind<VerificationMethod>>>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub did_document_metadata: Option<DocumentMetadata>,
}

/// Services are used to express ways of communicating with the DID subject or
/// associated entities. They can be any type of service the DID subject wants
/// to advertise, including decentralized identity management services for further
/// discovery, authentication, authorization, or interaction.
///
/// Service information is often service specific. For example, a reference to an
/// encrypted messaging service can detail how to initiate the encrypted link before
/// messaging begins.
///
/// Due to privacy concerns, revealing public information through services, such as
/// social media accounts, personal websites, and email addresses, is discouraged.
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct Service {
    /// A URI unique to the service.
    pub id: String,

    /// The service type. SHOULD be registered in the DID Specification Registries.
    #[serde(rename = "type")]
    pub type_: String,

    /// One or more endpoints for the service.
    #[allow(clippy::struct_field_names)]
    pub service_endpoint: Quota<Kind<Value>>,
}

/// A DID document can express verification methods, such as cryptographic public keys,
/// which can be used to authenticate or authorize interactions with the DID subject or
/// associated parties. For example, a cryptographic public key can be used as a
/// verification method with respect to a digital signature; in such usage, it verifies
/// that the signer could use the associated cryptographic private key. Verification
/// methods might take many parameters. An example of this is a set of five
/// cryptographic keys from which any three are required to contribute to a
/// cryptographic threshold signature.
///
/// MAY include additional properties which can be determined from the verification
/// method as registered in the
/// [DID Specification Registries](https://www.w3.org/TR/did-spec-registries/).
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct VerificationMethod {
    /// Only used when the verification method uses terms not defined in the containing
    /// document.
    #[serde(rename = "@context")]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub context: Option<Kind<Value>>,

    /// A DID that identifies the verification method.
    pub id: String,

    /// Kind references a verification method type. SHOULD be a registered type
    /// (in DID Specification Registries).
    #[serde(rename = "type")]
    pub type_: PublicKeyFormat,

    /// The DID of the controller of the verification method.
    pub controller: String,

    /// The public key material for the verification method.
    #[serde(flatten)]
    pub public_key: PublicKey,
}

/// Cryptographic curve type.
#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
pub enum PublicKeyFormat {
    /// Key is encoded as a Multibase.
    #[default]
    Multikey,

    /// ED2559 Verification Key, 2020 version
    Ed25519VerificationKey2020,

    /// X25519 Key Agreement Key, 2020 version
    X25519KeyAgreementKey2020,
    //
    // /// JsonWebKey2020
    // JsonWebKey2020,
}

// TODO: set context based on key format:
// - Ed25519VerificationKey2020	https://w3id.org/security/suites/ed25519-2020/v1
// - JsonWebKey2020	https://w3id.org/security/suites/jws-2020/v1

impl Display for PublicKeyFormat {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        match self {
            Self::Multikey => write!(f, "Multikey"),
            Self::Ed25519VerificationKey2020 => write!(f, "Ed25519VerificationKey2020"),
            Self::X25519KeyAgreementKey2020 => write!(f, "X25519KeyAgreementKey2020"),
        }
    }
}

/// Cryptographic curve type.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub enum PublicKey {
    /// Public key encoded as a Multibase.
    #[serde(rename = "publicKeyMultibase")]
    Multibase(String),

    /// Public key encoded as a JWK.
    #[serde(rename = "publicKeyJwk")]
    Jwk(PublicKeyJwk),
}

impl Default for PublicKey {
    fn default() -> Self {
        Self::Multibase(String::new())
    }
}

#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
#[allow(clippy::module_name_repetitions)]
pub struct DocumentMetadata {
    /// Timestamp of the Create operation.
    /// An XMLSCHEMA11-2 (RFC3339) e.g. 2010-01-01T19:23:24Z.
    created: DateTime<Utc>,

    /// Timestamp of the last Update operation. Omitted if an Update operation has
    /// never been performed. May be the same value as the `created` property when
    /// the difference between the two timestamps is less than one second.
    /// An XMLSCHEMA11-2 (RFC3339) e.g. 2010-01-01T19:23:24Z.
    #[serde(skip_serializing_if = "Option::is_none")]
    updated: Option<DateTime<Utc>>,

    /// MUST be set to true if the DID has been deactivated. Optional if the DID
    /// has not been deactivated.
    #[serde(skip_serializing_if = "Option::is_none")]
    deactivated: Option<bool>,

    /// May be set if the document version is not the latest. Indicates the timestamp
    /// of the next Update operation as an XMLSCHEMA11-2 (RFC3339).
    next_update: Option<DateTime<Utc>>,

    /// Used to indicate the version of the last Update operation. SHOULD be set.
    #[serde(skip_serializing_if = "Option::is_none")]
    version_id: Option<String>,

    /// MAY be set if the document version is not the latest. It indicates the version
    /// of the next Update operation.
    #[serde(skip_serializing_if = "Option::is_none")]
    next_version_id: Option<String>,

    /// Used when a DID method needs to define different forms of a DID that are
    /// logically equivalent. For example, when a DID takes one form prior to
    /// registration in a verifiable data registry and another form after such
    /// registration.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub equivalent_id: Option<Vec<String>>,

    /// Identical to the `equivalent_id` property except that it is a single value
    /// AND the DID is the canonical ID for the DID subject within the containing
    /// DID document.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub canonical_id: Option<String>,

    pub method: MethodMetadata,
}

#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct MethodMetadata {
    pub published: bool,
    pub recovery_commitment: String,
    pub update_commitment: String,
}
