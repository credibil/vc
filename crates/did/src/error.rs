/// DID resolution error codes
#[derive(thiserror::Error, Debug)]
pub enum Error {
    /// The DID method is not supported.
    #[error("methodNotSupported")]
    MethodNotSupported(String),

    /// The DID supplied to the DID resolution function does not conform to
    /// valid syntax.
    #[error("invalidDid")]
    InvalidDid(String),

    /// The DID resolver was unable to find the DID document resulting from
    /// this resolution request.
    #[error("notFound")]
    NotFound(String),

    /// The representation requested via the accept input metadata property is
    /// not supported by the DID method and/or DID resolver.
    #[error("representationNotSupported")]
    RepresentationNotSupported(String),

    /// The DID URL is invalid
    #[error("invalidDidUrl")]
    InvalidDidUrl(String),

    // ---- Creation Errors ----  //
    /// The byte length of raw public key does not match that expected for the
    /// associated multicodecValue.
    #[error("invalidPublicKeyLength")]
    InvalidPublicKeyLength(String),

    /// The public key is invalid
    #[error("invalidPublicKey")]
    InvalidPublicKey(String),

    /// Public key format is not known to the implementation.
    #[error("unsupportedPublicKeyType")]
    UnsupportedPublicKeyType(String),

    /// Other, unspecified errors.
    #[error(transparent)]
    Other(#[from] anyhow::Error),
}

impl Error {
    /// Returns the error code.
    #[must_use]
    pub fn code(&self) -> String {
        self.to_string()
    }

    /// Returns the associated error message.
    #[must_use]
    pub fn message(&self) -> String {
        match self {
            Self::MethodNotSupported(msg)
            | Self::InvalidDid(msg)
            | Self::NotFound(msg)
            | Self::InvalidDidUrl(msg)
            | Self::RepresentationNotSupported(msg)
            | Self::InvalidPublicKeyLength(msg)
            | Self::InvalidPublicKey(msg)
            | Self::UnsupportedPublicKeyType(msg) => msg.clone(),
            Self::Other(err) => err.to_string(),
        }
    }
}

// impl From<anyhow::Error> for Error {
//     fn from(err: anyhow::Error) -> Self {
//         Self
//     }
// }

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn error_code() {
        let err = Error::MethodNotSupported("Method not supported".into());
        assert_eq!(err.message(), "Method not supported");
    }
}
