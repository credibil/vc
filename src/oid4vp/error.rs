//! # `OpenID` Errors
//!
//! This module defines errors for `OpenID` for Verifiable Credential Issuance
//! and Verifiable Presentations.

// TODO: add support for "client-state" in error responses.
// TODO: use custom serialisation for Err enum.

use serde::{Deserialize, Serialize, Serializer};
use thiserror::Error;

use crate::core::urlencode;

/// `OpenID` error codes for  for Verifiable Credential Issuance and
/// Presentation.
#[derive(Error, Debug, Deserialize)]
#[allow(clippy::enum_variant_names)]
pub enum Error {
    /// The request is missing a required parameter, includes an unsupported
    /// parameter value, repeats a parameter, includes multiple credentials,
    /// utilizes more than one mechanism for authenticating the client, or is
    /// otherwise malformed.
    #[error(r#"{{"error": "invalid_request", "error_description": "{0}"}}"#)]
    InvalidRequest(String),

    /// The authorization server does not support obtaining an authorization
    /// code using this method.
    #[error(r#"{{"error": "unsupported_response_type", "error_description": "{0}"}}"#)]
    UnsupportedResponseType(String),

    /// The authorization server encountered an unexpected condition that
    /// prevented it from fulfilling the request.
    #[error(r#"{{"error": "server_error", "error_description": "{0}"}}"#)]
    ServerError(String),

    /// The Wallet does not support any of the formats requested by the
    /// Verifier, such as those included in the `vp_format_supported` registration
    /// parameter.
    #[error(r#"{{"error": "vp_formats_not_supported", "error_description": "{0}"}}"#)]
    VpFormatsNotSupported(String),

    /// The Presentation Definition URL cannot be reached.
    #[error(r#"{{"error": "invalid_presentation_definition_uri", "error_description": "{0}"}}"#)]
    InvalidPresentationDefinitionUri(String),

    /// The Presentation Definition URL can be reached, but the specified
    /// `presentation_definition` cannot be found at the URL.
    #[error(
        r#"{{"error": "invalid_presentation_definition_reference", "error_description": "{0}"}}"#
    )]
    InvalidPresentationDefinitionReference(String),

    /// The Wallet appears to be unavailable and therefore unable to respond to
    /// the request.
    /// Use when the User Agent cannot invoke the Wallet and another component
    /// receives the request while the End-User wishes to continue the journey
    /// on the Verifier website.
    #[error(r#"{{"error": "wallet_unavailable", "error_description": "{0}"}}"#)]
    WalletUnavailable(String),
}

/// Error response for `OpenID` for Verifiable Credentials.
#[allow(clippy::module_name_repetitions)]
#[derive(Deserialize, Serialize)]
pub struct OidError {
    /// Error code.
    pub error: String,

    /// Error description.
    pub error_description: String,

    /// Optional client-state parameter.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub state: Option<String>,

    /// A fresh `c_nonce` to use when retrying Proof submission.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub c_nonce: Option<String>,

    /// The expiry time of the `c_nonce`.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub c_nonce_expires_in: Option<i64>,
}

impl Serialize for Error {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        use serde::ser::Error as SerdeError;

        let Ok(error) = serde_json::from_str::<OidError>(&self.to_string()) else {
            return Err(SerdeError::custom("issue deserializing Err"));
        };
        error.serialize(serializer)
    }
}

impl Error {
    /// Transfrom error to `OpenID` compatible json format.
    #[must_use]
    pub fn to_json(&self) -> serde_json::Value {
        serde_json::from_str(&self.to_string()).unwrap_or_default()
    }

    /// Transfrom error to `OpenID` compatible query string format.
    /// Does not include `c_nonce` as this is not required for in query
    /// string responses.
    #[must_use]
    pub fn to_querystring(&self) -> String {
        urlencode::to_string(&self).unwrap_or_default()
    }
}

#[cfg(test)]
mod test {
    use serde_json::{Value, json};

    use super::*;

    // Test that error details are retuned as json.
    #[test]
    fn err_json() {
        let err = Error::InvalidRequest("bad request".to_string());
        let ser: Value = serde_json::from_str(&err.to_string()).unwrap();
        assert_eq!(ser, json!({"error":"invalid_request", "error_description": "bad request"}));
    }

    // Test that the error details are returned as an http query string.
    #[test]
    fn err_querystring() {
        let err = Error::InvalidRequest("Invalid request description".to_string());
        let ser = urlencode::to_string(&err).unwrap();
        assert_eq!(ser, "error=invalid_request&error_description=Invalid%20request%20description");
    }

    // Test that the error details are returned as an http query string.
    #[test]
    fn err_serialize() {
        let err = Error::InvalidRequest("bad request".to_string());
        let ser = serde_json::to_value(&err).unwrap();
        assert_eq!(ser, json!({"error":"invalid_request", "error_description": "bad request"}));
    }
}
