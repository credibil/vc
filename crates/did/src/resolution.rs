//! # DID Resolver
//!
//! This crate provides a DID Resolver trait and a set of default
//! implementations for resolving DIDs.
//!
//! See [DID resolution](https://www.w3.org/TR/did-core/#did-resolution) fpr more.

use std::collections::HashMap;

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::document::{Document, DocumentMetadata, Service, VerificationMethod};
use crate::error::Error;
use crate::{key, web, DidResolver};

/// Resolve a DID to a DID document.
///
/// The [DID resolution](https://www.w3.org/TR/did-core/#did-resolution) functions
/// resolve a DID into a DID document by using the "Read" operation of the
/// applicable DID method.
///
/// Caveats:
/// - No JSON-LD Processing, however, valid JSON-LD is returned.
/// - Ignores accept header.
/// - Only returns application/did+ld+json.
/// - did:key support for ed25519
/// - did:web support for .well-known and path based DIDs.
///
/// # Errors
///
/// Returns a [DID resolution](https://www.w3.org/TR/did-core/#did-resolution-metadata)
/// error as specified.
pub async fn resolve(
    did: &str, opts: Option<Options>, resolver: &impl DidResolver,
) -> crate::Result<Resolved> {
    // use DID-specific resolver
    let method = did.split(':').nth(1).unwrap_or_default();

    let result = match method {
        "key" => key::DidKey::resolve(did, opts, resolver),
        "web" => web::DidWeb::resolve(did, opts, resolver).await,
        _ => Err(Error::MethodNotSupported(format!("{method} is not supported"))),
    };

    if let Err(e) = result {
        return Ok(Resolved {
            metadata: Metadata {
                error: Some(e.to_string()),
                error_message: Some(e.message()),
                content_type: ContentType::DidLdJson,
                ..Metadata::default()
            },
            ..Resolved::default()
        });
    };

    result
}

/// Dereference a DID URL into a resource.
///
/// # Errors
pub async fn dereference(
    did_url: &str, opts: Option<Options>, resolver: &impl DidResolver,
) -> crate::Result<Dereferenced> {
    // extract DID from DID URL
    let url = url::Url::parse(did_url)
        .map_err(|e| Error::InvalidDidUrl(format!("issue parsing URL: {e}")))?;
    let did = format!("did:{}", url.path());

    // resolve DID document
    let method = did_url.split(':').nth(1).unwrap_or_default();
    let resolution = match method {
        "key" => key::DidKey::resolve(&did, opts, resolver)?,
        "web" => web::DidWeb::resolve(&did, opts, resolver).await?,
        _ => return Err(Error::MethodNotSupported(format!("{method} is not supported"))),
    };

    let Some(document) = resolution.document else {
        return Err(Error::InvalidDid("Unable to resolve DID document".into()));
    };

    // process document to dereference DID URL for requested resource
    let Some(verifcation_methods) = document.verification_method else {
        return Err(Error::NotFound("verification method missing".into()));
    };

    // for now we assume the DID URL is the ID of the verification method
    // e.g. did:web:demo.credibil.io#key-0
    let Some(vm) = verifcation_methods.iter().find(|vm| vm.id == did_url) else {
        return Err(Error::NotFound("verification method not found".into()));
    };

    Ok(Dereferenced {
        metadata: Metadata {
            content_type: ContentType::DidLdJson,
            ..Metadata::default()
        },
        content_stream: Some(Resource::VerificationMethod(vm.clone())),
        content_metadata: Some(ContentMetadata {
            document_metadata: resolution.document_metadata,
        }),
    })
}

/// Used to pass addtional values to a `resolve` and `dereference` methods. Any
/// properties used should be registered in the DID Specification Registries.
///
/// The `accept` property is common to all resolver implementations. It is used
/// by users to specify the Media Type when calling the `resolve_representation`
/// method. For example:
///
/// ```json
/// {
///    "accept": "application/did+ld+json"
/// }
/// ```
#[derive(Clone, Debug, Default, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Options {
    /// [`accept`](https://www.w3.org/TR/did-spec-registries/#accept) resolution option.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub accept: Option<ContentType>,

    // pub public_key_format: Option<String>,
    /// Additional options.
    #[serde(flatten)]
    pub additional: Option<HashMap<String, Metadata>>,
}

/// The DID URL syntax supports parameters in the URL query component. Adding a
/// DID parameter to a DID URL means the parameter becomes part of the
/// identifier for a resource.
#[derive(Clone, Debug, Default, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Parameters {
    /// Identifies a service from the DID document by service's ID.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub service: Option<String>,

    /// A relative URI reference that identifies a resource at a service
    /// endpoint, which is selected from a DID document by using the service
    /// parameter. MUST use URL encoding if set.
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(alias = "relative-ref")]
    pub relative_ref: Option<String>,

    /// Identifies a specific version of a DID document to be resolved (the
    /// version ID could be sequential, or a UUID, or method-specific).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub version_id: Option<String>,

    /// Identifies a version timestamp of a DID document to be resolved. That
    /// is, the DID document that was valid for a DID at a certain time.
    /// An XML datetime value [XMLSCHEMA11-2] normalized to UTC 00:00:00 without
    /// sub-second decimal precision. For example: 2020-12-20T19:17:47Z.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub version_time: Option<String>,

    /// A resource hash of the DID document to add integrity protection, as
    /// specified in [HASHLINK]. This parameter is non-normative.
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "hl")]
    pub hashlink: Option<String>,

    /// Additional parameters.
    #[serde(flatten)]
    pub additional: Option<HashMap<String, Value>>,
}

/// Returned by `resolve` DID methods.
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct Resolved {
    /// The DID resolution context.
    #[serde(rename = "@context")]
    pub context: String,

    /// Resolution metadata.
    pub metadata: Metadata,

    /// The DID document.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub document: Option<Document>,

    /// DID document metadata.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub document_metadata: Option<DocumentMetadata>,
}

/// `Dereferenced` contains the result of dereferencing a DID URL.
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct Dereferenced {
    /// A metadata structure consisting of values relating to the results of the
    /// DID URL dereferencing process. MUST NOT be empty in the case of an
    /// error.
    pub metadata: Metadata,

    /// The dereferenced resource corresponding to the DID URL. MUST be empty if
    /// dereferencing was unsuccessful. MUST be empty if dereferencing is
    /// unsuccessful.
    pub content_stream: Option<Resource>,

    /// Metadata about the `content_stream`. If `content_stream` is a DID
    /// document, this MUST be `DidDocumentMetadata`. If dereferencing is
    /// unsuccessful, MUST be empty.
    pub content_metadata: Option<ContentMetadata>,
}

/// Resource represents the DID document resource returned as a result of DID
/// dereferencing. The resource is a DID document or a subset of a DID document.
#[allow(clippy::large_enum_variant)]
#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
pub enum Resource {
    ///  DID `Document` resource.
    Document(Document),

    /// `VerificationMethod` resource.
    VerificationMethod(VerificationMethod),

    /// `Service` resource.
    Service(Service),
}

impl Default for Resource {
    fn default() -> Self {
        Self::VerificationMethod(VerificationMethod::default())
    }
}

/// DID document metadata.
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct Metadata {
    /// The Media Type of the returned resource.
    pub content_type: ContentType,

    /// The error code from the dereferencing process, if applicable.
    /// Values of this field SHOULD be registered in the DID Specification
    /// Registries. Common values are `invalid_did_url` and `not_found`.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,

    /// A human-readable explanation of the error.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error_message: Option<String>,

    /// Additional information about the resolution or dereferencing process.
    #[serde(flatten)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub additional: Option<Value>,
}

/// The Media Type of the returned resource.
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
pub enum ContentType {
    /// JSON-LD representation of a DID document.
    #[default]
    #[serde(rename = "application/did+ld+json")]
    DidLdJson,
    //
    // /// The JSON-LD Media Type.
    // #[serde(rename = "application/ld+json")]
    // LdJson,
}

/// Metadata about the `content_stream`. If `content_stream` is a DID document,
/// this MUST be `DidDocumentMetadata`. If dereferencing is unsuccessful, MUST
/// be empty.
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ContentMetadata {
    /// The DID document metadata.
    #[serde(flatten)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub document_metadata: Option<DocumentMetadata>,
}

#[cfg(test)]
mod test {
    use anyhow::anyhow;
    use insta::assert_json_snapshot as assert_snapshot;

    use super::*;

    struct MockResolver;
    impl DidResolver for MockResolver {
        async fn resolve(&self, _url: &str) -> anyhow::Result<Document> {
            serde_json::from_slice(include_bytes!("web/did-ecdsa.json"))
                .map_err(|e| anyhow!("issue deserializing document: {e}"))
        }
    }

    #[test]
    fn error_code() {
        let err = Error::MethodNotSupported("Method not supported".into());
        assert_eq!(err.message(), "Method not supported");
    }

    #[tokio::test]
    async fn deref_web() {
        const DID_URL: &str = "did:web:demo.credibil.io#key-0";

        let dereferenced =
            dereference(DID_URL, None, &MockResolver).await.expect("should dereference");
        assert_snapshot!("deref_web", dereferenced);
    }

    #[tokio::test]
    async fn deref_key() {
        const DID_URL: &str = "did:key:z6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK#z6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK";

        let dereferenced =
            dereference(DID_URL, None, &MockResolver).await.expect("should dereference");
        assert_snapshot!("deref_key", dereferenced);
    }
}
