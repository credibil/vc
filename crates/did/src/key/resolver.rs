//! # DID Key Resolver
//!
//! The `did:key` method is a DID method for static cryptographic keys. At its
//! core, it is based on expanding a cryptographic public key into a DID
//! Document.
//!
//! See:
//!
//! - <https://w3c-ccg.github.io/did-method-key>
//! - <https://w3c.github.io/did-resolution>

use std::sync::LazyLock;

use regex::Regex;
use serde_json::json;

use super::DidKey;
use crate::document::CreateOptions;
use crate::error::Error;
use crate::resolution::{ContentType, Metadata, Options, Resolved};
use crate::DidResolver;

const ED25519_CODEC: [u8; 2] = [0xed, 0x01];
static DID_REGEX: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new("^did:key:(?<identifier>z[a-km-zA-HJ-NP-Z1-9]+)$").expect("should compile")
});

impl DidKey {
    pub fn resolve(did: &str, _: Option<Options>, _: &impl DidResolver) -> crate::Result<Resolved> {
        // check DID is valid AND extract key
        let Some(caps) = DID_REGEX.captures(did) else {
            return Err(Error::InvalidDid("DID is not a valid did:key".into()));
        };
        let multikey = &caps["identifier"];

        // decode the the DID key
        let (_, key_bytes) = multibase::decode(multikey)
            .map_err(|e| Error::InvalidDid(format!("issue decoding key: {e}")))?;
        if key_bytes.len() - 2 != 32 {
            return Err(Error::InvalidDid("invalid key length".into()));
        }
        if key_bytes[0..2] != ED25519_CODEC {
            return Err(Error::InvalidDid("unsupported signature".into()));
        }

        // per the spec, use the create operation to generate a DID document
        let options = CreateOptions {
            enable_encryption_key_derivation: true,
            ..CreateOptions::default()
        };

        let document =
            Self::create(&key_bytes[2..], options).map_err(|e| Error::InvalidDid(e.to_string()))?;

        Ok(Resolved {
            context: "https://w3id.org/did-resolution/v1".into(),
            metadata: Metadata {
                content_type: ContentType::DidLdJson,
                additional: Some(json!({
                    "pattern": "^did:key:z[a-km-zA-HJ-NP-Z1-9]+$",
                    "did": {
                        "didString": did,
                        "methodSpecificId": did[8..],
                        "method": "key"
                    }
                })),
                ..Metadata::default()
            },
            document: Some(document),
            ..Resolved::default()
        })
    }
}

#[cfg(test)]
mod test {
    use insta::assert_json_snapshot as assert_snapshot;

    use super::*;
    use crate::document::Document;

    const DID: &str = "did:key:z6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK";

    struct MockResolver;
    impl DidResolver for MockResolver {
        async fn resolve(&self, _url: &str) -> anyhow::Result<Document> {
            Ok(Document::default())
        }
    }

    #[tokio::test]
    async fn resolve() {
        let resolved = DidKey::resolve(DID, None, &MockResolver).expect("should resolve");
        assert_snapshot!("resolved", resolved);
    }
}
