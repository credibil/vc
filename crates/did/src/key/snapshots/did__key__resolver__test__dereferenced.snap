---
source: crates/did/src/key/resolver.rs
assertion_line: 135
expression: dereferenced
---
{
  "metadata": {
    "contentType": "application/did+ld+json"
  },
  "contentStream": {
    "VerificationMethod": {
      "id": "did:key:z6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK#z6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK",
      "type": "Multikey",
      "controller": "did:key:z6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK",
      "publicKeyMultibase": "z6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK"
    }
  },
  "contentMetadata": {}
}
